File: 프로그래머스/Level3/깊이너비 우선 탐색(DFS BFS) - 단어변환/단어변환.swift
import Foundation

func solution(_ begin:String, _ target:String, _ words:[String]) -> Int {
    if !words.contains(target) {
        return 0
    }
    
    var curStep: [String] = []
    var nextStep: [String] = []
    var visited: Set<String> = []
    
    curStep.reserveCapacity(words.count)
    
    var curWord = begin
    var depth: Int = 0  
    while visited.count <= words.count {
        // 현재 단어가 target이면 리턴
        if curWord == target {
            return depth
        }
        // 현재 단어와 한글자 차이인 단어들을 nextStep에 넣기
        for word in words {
            if visited.contains(word) {continue}
            if compareDiff(curWord, word) == 1 {
                nextStep.append(word)
            }
        }        
        // 현재 step을 다 순회 했으면 다음 step으로 이동
        if curStep.isEmpty {
            curStep = nextStep
            nextStep = []
            depth += 1
        }
        curWord = curStep.removeFirst()
        visited.insert(curWord)
    }
    
    return 0
}

func compareDiff(_ str1: String, _ str2: String) -> Int {
    var count = 0
    for i in 0..<str1.count {
        if str1[str1.index(str1.startIndex, offsetBy: i)] != str2[str2.index(str2.startIndex, offsetBy: i)] {
            count += 1
        }
    }
    return count
}

File: 프로그래머스/Level3/깊이너비 우선 탐색(DFS BFS) - 단어변환/단어변환.py
def checkDiff(word1, word2):
    if len(word1) != len(word1):
        return -1
    
    count = 0
    for i in range(len(word1)):
        if word1[i] != word2[i]:
            count += 1
            
    return count

def bfs(graph, begin, target):
    queue = [[begin,0]] # 방문할 노드의 문자(값)와 깊이를 저장
    visited = []        # 방문한 노드의 문자(값)만 저장
    
    while queue:
        # 큐에서 값을 하나 꺼낸다
        node = queue[0]
        del queue[0]
        
        word = node[0]
        depth = node[1]
        
        # 꺼낸 값이 target과 같다.
        if word == target:
            return depth
        
        # 방문한 적이 없는 node다.        
        if word not in visited:
            visited.append(word) # 방문 기록에 추가
            
            # 모든 자식노드를 깊이+1과 zip()하여 큐에 추가(extend)한다.
            queue.extend(zip(graph[word], [depth+1]*len(graph[word]))) 
    
    # 모든 queue를 다 돌 동안 target을 만나지 못했다
    return 0
 
def solution(begin, target, words):
    # 명백한 반환 조건
    if target not in words:
        return 0
    
    # 그래프 만들기
    graph = {}
    
    ## begin : word
    for word in words:
        if checkDiff(begin, word) == 1:
            if word in graph:
                graph[word].append(begin)
            else:
                graph[word] = [begin]
            if begin in graph:
                graph[begin].append(word)
            else:
                graph[begin] = [word]
                
    ## word : word
    for i in range(len(words)):
        for j in range(i+1, len(words)):
            if checkDiff(words[i], words[j]) == 1:
                if words[i] in graph:
                    graph[words[i]].append(words[j])
                else:
                    graph[words[i]] = [words[j]]
                if words[j] in graph:
                    graph[words[j]].append(words[i])
                else:
                    graph[words[j]] = [words[i]]
        
    return bfs(graph, begin, target)